### ISSUE TITLE
[Feature request] add terminal_mappings

### ISSUE THREAD

**Issue by briandipalma** [akinsho/toggleterm.nvim#122]
Means you can use the mapping to open and close a terminal when `insert_mappings` is set to false. My mapping uses my leader key (space) which was causing issues in insert mode. But if the mapping could work in terminal mode it would be great. It's a tiny change, happy to raise a PR if it makes sense?

**Comment by akinsho** [comment]
Hi @briandipalma, so I'm not sure if I understand the problem you're describing. Can you rephrase this problem or clarify further?

**Comment by briandipalma** [comment]
My open terminal mapping is `spacebar (leader) tt`, so in normal mode there is a pause when I hit spacebar. I had to set `insert_mappings` to false to stop it. I like toggling the terminal when in terminal mode so I added a terminal mapping to do that. I was thinking it should probably be built in to the plugin for cases like mine though.

**Comment by akinsho** [comment]
@briandipalma if I understand this correctly the particular issue is with the fact that you use spacebar as a leader key since on my end I'm able to toggle the terminal with my open key even in insert mode? I'm not sure a specific binding to circumvent this edge case is required if it can be fixed via a config change and is specific to the open key a user chooses.
I might have misunderstood this issue still though

**Comment by akinsho** [comment]
Although one potential area that needs fixing is that the `terminal` mapping is also set only if `insert` mappings is true
https://github.com/akinsho/toggleterm.nvim/blob/f725433e59422d0514ffa2582c534da965b0e44f/lua/toggleterm/terminal.lua#L82
🤔 I don't like the idea of yet another configuration option called like `terminal` mappings though. I guess if I were going to allow any change it would be to just always set a terminal mapping but feel like the immediate fallout of that will be that breaks someone else's setup

**Comment by briandipalma** [comment]
I was thinking more of another if condition here: https://github.com/akinsho/toggleterm.nvim/blob/f725433e59422d0514ffa2582c534da965b0e44f/lua/toggleterm.lua#L40

The issue is that, if someone uses a leader based mapping to toggle the terminal, whenever the leader key is hit in insert mode Neovim pauses to see if the complete toggle terminal key combination is entered. This pause mess up typing flow. Maybe I'll have a think about my mapping and change it to something other than a leader based one.

**Comment by akinsho** [comment]
@briandipalma this has always been a pitfall with insert mappings and nvim I'm not sure there's much I can do here apart from a user changing their mapping or not using insert mappings. Any insert mappings in any context in nvim come with the risk of triggering the timeoutlen 🤷🏿‍♂️ which is why it can be turned off.

**Comment by briandipalma** [comment]
Yeah fair enough, I'll experiment with some other keybindings instead :+1:

**Comment by MageJohn** [comment]
@akinsho could this issue be re-opened? I'd like the option as described in the issue title, specifically a `terminal_mappings` option separate from the `insert_mappings` option. In my case I don't want my terminal mapping in insert mode when I'm typing, but I do want it in terminal mode so I can easily toggle the terminal open and closed. 

Also, having the option named `insert_mappings` be in control of terminal mode mappings feels like surprising behaviour to me. It would feel more consistent to me if they were separate options.

If you agree I'm happy to open a PR!

**Comment by akinsho** [comment]
@MageJohn a PR to add this would be great I'm not opposed to being more granular here as long as by default it behaves the same way as it currently does so it doesn't break for existing users

**Comment by MageJohn** [comment]
I've opened a PR! The default behaves the same, though people who have disabled `insert_mappings` already will now have terminal mode mappings by default. Hopefully the need to update will be obvious from the commit message.

**Comment by akinsho** [comment]
Fixed by #149 thanks again for the contribution @MageJohn
